### PR TITLE
[WIP] Complex values support

### PR DESCRIPTION
**Pull request by smartass101**
Hi,

first of all, thank you for creating this package, it is nearly exactly what I've been looking for. 
However, for my application I need to demix complex-valued signals which are somewhat non-stationary. The [original UWEDGE_C algorithm](http://si.utia.cas.cz/downloadPT.htm) obtained from the original authors works quite well for demixing the complex signals, but I'd like to use your partitioning approach to account for the non-stationarity.

I've attempted to modify the uwedge function to support complex values, but so far it fails to converge (or perhaps a different convergence check is in order).
I've been mostly guided by the MATLAB source code obtained from the original authors of UWEDGE{,_C} and by the following excerpt from [their article](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=4671095):

>When the
target-matrices are complex-valued and Hermitian, and the
diagonal source matrices are real-valued, extension of our algorithms for finding complex-valued mixing/demixing matrices is
straightforward, and merely involves substitution of all “transpose” operations with “conjugate transpose” (thus, the same
Matlab code can be readily applied in such cases).In the fully
complex framework, in which the diagonal matrices may also
be complex-valued (and the target-matrices non-Hermitian),
adaptation of our algorithms is more involved, but possible

I also had to disable the sklearn-style matrix checks, because they do not support complex matrices.

To test my implementation I also added simple test inspired by the Simulations section in the original UWEDGE paper. Unfortunately, **even with the unmodified uwedge function the tests fail** for real values.

Any help would be greatly appreciated!

### CODE REVIEW

**Comment by sweichwald** [comment]
Hi, thanks for the feedback and contribution!

Note that the current test fails since we only expect the two matrices `V` and `V_` to be close up to permutation and sign of the rows. The test needs to take this into account. (Also note that `allclose` by default has 0 absolute tolerance and only relative tolerance, which is a problem when comparing against 0's.)

As an example, when only considering orthogonal `A`'s like in the provided test scenario, the following test would work for real matrices; it compares the row-sorted and -signed product of `V` and `V_.T` to the identity matrix.
```python
    # compare VV_' against the identity up to permutation/sign of rows
    VVt = V.dot(V_.T)
    VVtsort = VVt[np.argsort(np.argmax(np.abs(VVt), axis=1)), :]
    VVtsortsign = np.sign(np.diagonal(VVtsort)) * VVtsort
    np.testing.assert_allclose(VVtsortsign, np.eye(d), rtol=0, atol=1e-9)
```
More generally, something like the MD index (https://link.springer.com/chapter/10.1007/978-3-642-15995-4_29 or Section 2.4. in http://arxiv.org/abs/1806.01094) could be used for testing
```python
from scipy.optimize import linear_sum_assignment


def MDindex(A, Vhat):
    d = np.shape(A)[0]
    G = Vhat.dot(A)
    Gtilde = (G**2) / ((G**2).sum(axis=1)).reshape((d, 1))
    costmat = 1 - 2 * Gtilde + np.tile((Gtilde**2).sum(axis=0), d).reshape((d, d))
    row_ind, col_ind = linear_sum_assignment(costmat)
    md = np.sqrt(d - np.sum(np.diag(Gtilde[row_ind, col_ind]))) / np.sqrt(d - 1)
    return md
```

Maybe this serves as a good starting point for the complex extension?

Some notes:
* what is the best way to deal with potentially complex inputs in sklearn (sklearn tests)
* add pytest to requirements-test.txt

Cheers, @NiklasPfister and @sweichwald.

PS. If you don't mind sharing we'd be really curious to hear what you're using coroICA for ;-)

**Comment by smartass101** [comment]
Hi, thank you for getting back to me, I appreciate your interest.

Thank you for pointing out the MD index. I've put it into `coroica.utils.md_index` and used `abs(G)**2` to support complex values. Now al least the real-valued` tests pass :smile: So at least my complex extensions din't break anything and now I can investigate the complex case.

## My use-case

In the field of experimental tokamak physics  one often measures magnetic fluctuations via an array of magnetic pickup coils due to magnetohydrodynamic modes rotating relative to the array. Therefore, for each coil `j` one measures a superposition of several modes in the form (not necessarily completely harmonic) `\dot B_j(t) = \sum_i A_i \cos(\omega_i t + m_i \theta_j^* + n_i \phi_j)`. The goal usually  is to identify and analyze the evolution of `A_i(t), m_i, n_i, \omega_i(t)`. Since each coil has a different phase for the signals depending on the mode structure, ICA alone does not work. A classical result of [Nardone](https://iopscience.iop.org/article/10.1088/0741-3335/34/9/001) was to use SVD to identify orthogonal `cos` and `sin` bases both in space and time and then `\cos(\omega t + \phi) = a \cos(\omega t) + b  \sin(\omega t)`. Unfortunately, this often fails when the modes have similar structure and/or frequency as is the case in some of our experimental data.

My approach is to try proper SOBI-style ICA to separate modes with different (even quite close) frequencies and any poloidal structure, i.e. orthogonalize only in time, not space . My main idea to enable this is to use the Hilbert transform (which is linear) to get the analytic signal, which incorporates the phase as well as the amplitude and is admitted in the [SOBI model](http://www.bsp.brain.riken.jp/ICApub/A97-434.pdf) which is defined for complex signals in general. The UWEDGE_C algorithm seems to work quite well for some of our experimental data, but I hope to extend coroICA to account for non-stationarity.

**Comment by sweichwald** [comment]
Great, thanks!

Let us know how it goes :-)

PS. I seem unable to reproduce the problem with `check_array` and `check_X_y`, i.e. the tests work for me. What problem do you run into when you leave the checks in uwedgeica uncommented? Are the `dtype` parameters of `check_array` and `check_X_y` of help for you?

**Comment by sweichwald** [comment]
Great, thanks!

Let us know how it goes :-)

PS. I seem unable to reproduce the problem with `check_array` and `check_X_y`, i.e. the tests work for me. What problem do you run into when you leave the checks in uwedgeica uncommented? Are the `dtype` parameters of `check_array` and `check_X_y` of help to you?

**Comment by smartass101** [comment]
Looking at the [source code of `check_array` here](https://github.com/scikit-learn/scikit-learn/blob/7b136e9/sklearn/utils/validation.py#L524) my understanding  is that complex data is not supported in general, regardless of what I supply as the `dtype` argument.

**Comment by smartass101** [comment]
I finally got it to converge (for the smaller cases) to pretty much the same result like the UWEDGE_C algorithm, but the `md_index` is quite high, above 0.7. For truly inverse complex matrices the `md_index` (with `abs(G)**2` gives 0 and for perturbed above 0, so I hope this metric works as it should, but maybe for complex numbers it is different. It seems to give mostly reasonable result for real data, but I'm not confident with the md_index so high.

**Comment by sweichwald** [comment]
Hi @smartass101,
@NiklasPfister and myself just pushed a commit implementing complex value support.
Please let us know and reopen this issue in case you encounter problems with complex value support in `coroICA==0.1.21`.
Thanks!

PS. There was a `.conj()` missing in `sim_samples`, but otherwise we are using the tests you were suggesting. Please let us know in case there is a particular reason to prefer `eig` over `eigh` here; we left it unchanged as the latter is for Hermitian/symmetric matrices. Otherwise we added a few `.conj()` and `np.abs` much like you did.

**Comment by smartass101** [comment]
Thank you for this pleasant and unexpected surprise! Hope my code attempts were of some use and inspiration.

Regarding the `eig` usage, I had the feeling that `Rx0` may not be always Hermitian, because if I understood the code correctly, if one uses `instantcov=False`, then `Rx0` might end up being `Rx1` as per [this line](https://github.com/sweichwald/coroICA-python/blob/cc7818b9aab6ca2d0ebc84d54fe1df84e79b0e9e/coroica/uwedge.py#L31). But I suppose we should discuss this in a different issue.

How did you solve the `check_array` from `sklearn` issues?

**Comment by sweichwald** [comment]
Yes, thanks again for your pull request and your contribution!

The issue with sklearn checks was just resolved (cf. #2). (I did not encounter it since I was still on an older sklearn version.) There is now an option `skip_sklearn_checks`, which disables the checks alltogether. Use with caution ;-)

The way `uwedge` is called from within `CoroICA.fit` or `UwedgeICA.fit` the use of `eigh` should not be a problem since we always provide `Rx0 = np.cov(X)`. When calling `uwedge` separately, however, it may indeed happen that `eigh` is called for non-Hermitian matrices.
I prefer to use `eigh` if the matrix is known to be symmetric (faster, numerically more stable, eigenvalues sorted). Maybe there should be an option that allows the user to switch to `eig` or a check for symmetry.
